Improve todo widget usability

pratik-pc opened the pull request:
Introduces edit option for todo options.

Current design-
* edit icon adjacent to each todo item.
* only todo list author can edit items (other user can't edit todo items added by them)
* can toggle input fields for multiple todo items
* edit item input state persists on rerenders

Fixes part of zulip#20213
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Screenshots and screen captures:</summary>
<img width="707" height="250" alt="image" src="https://github.com/user-attachments/assets/0a439533-4e4e-45f6-b0bf-c77c3055f26f" />
<img width="707" height="250" alt="image" src="https://github.com/user-attachments/assets/b72f57c3-13f1-4a04-abaf-5b66802f531d" />

Error message on duplicate names
<img width="784" height="260" alt="image" src="https://github.com/user-attachments/assets/e0e018f8-f3e2-4b5e-9fdd-ac91dc3191e8" />

[Screencast from 2026-03-31 02-40-48.webm](https://github.com/user-attachments/assets/9221b7f7-09ce-485c-8927-8ee6382510ca)

|Before 1st commit|After 1st commit|
|-|-|
|<img width="638" height="260" alt="Screenshot from 2026-06-05 23-21-06" src="https://github.com/user-attachments/assets/f9d2f32b-2a8f-4740-b64b-600e1f38e899" />|<img width="638" height="260" alt="Screenshot from 2026-06-05 23-17-48" src="https://github.com/user-attachments/assets/da9f80aa-1ce5-42d2-827a-0e0fd7f9bc65" />|

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Cool! The pencil buttons look like a lot... I think we'll want them just on hover.

pratik-pc commented:
> Cool! The pencil buttons look like a lot... I think we'll want them just on hover.

This is how they look like now
![Screenshot from 2025-03-04 01-03-42](https://github.com/user-attachments/assets/f4d94804-4069-4501-b789-4ce0779a9e41)

This is the area where pencil button will appear on hover
![Screenshot from 2025-03-04 01-03-34](https://github.com/user-attachments/assets/b60d93e4-42b5-47c3-958d-2abe37b58424)

alya commented:
@amanagr Are you up for reviewing this one?

amanagr commented:
![image](https://github.com/user-attachments/assets/21638c0e-c6dd-4181-bb31-83627a746863)

The focus box is not visible on right side of button.

pratik-pc commented:
I think the backend tests are failing in all PRs, but anyway, I’ve added a margin for the outline.

IEC61960 reviewed (commented):
[vossa15 Thai National ID Card
1 4102 00080 53 3
Identification Number
Mr. Methi
Last name
Yaowala
unoiuf 1 d.A. 2532
Date of Birth 1 Aug. 1989
5.09557%
45-9.0. 2567
31 n.A. 2576
711/07И89078
15 Oct. 2024
31 Jul. 2033
Date of Issue
19744774000105
Date of Expiry
4102-11-10151540
บริษัท อมตะ ฟาซิลิตี้ เซอร์วิส จำกัด
700/2 หมู่ 1 ถนนบางนา-ตราด
ตำบลคลองตำหรุ อำเภอเมืองชลบุรี
จังหวัดชลบุรี 20000 (ประเทศไทย)](url)

amanagr commented:
Thanks @pratik-pc! Changes look good to me and in manual testing too. Can you fix the CI?

pratik-pc commented:
CI failing again for some PR I guess

pratik-pc commented:
Checks have passed now

alya commented:
Currently, when you edit a task, it disappears for a bit and then reappears. Can we just leave it in place until we're ready to update it?

alya commented:
> can toggle input fields for multiple todo items but submitting closes all input fields

I don't think that behavior makes sense: it can lead someone astray if they start editing several tasks in parallel. If we want to allow just one update at a time, then we should not even show edit buttons on other tasks while an edit is in progress.

pratik-pc commented:
> Currently, when you edit a task, it disappears for a bit and then reappears, which feels rather confusing. Can we just show it as updated right away, even if the server is still confirming?

Fixed this; I thought I was already doing this but there was bit of a formatting error.

> I don't think that behavior makes sense: it can lead someone astray if they start editing several tasks in parallel. If we want to allow just one update at a time, then we should not even show edit buttons on other tasks while an edit is in progress.

I reworked whole `todo_widget` rendering logic; earlier entire todo list was rerendered on new updates which would lead to input field closing automatically. Now todo lists are updated in-place without rerendering and input fields aren't affected by server events. Also, one should be able to edit multiple items and input fields won't close.

I could restrict it to one update at a time if we want to stick with that

alya commented:
The approach you describe where the rest of the fields stay open sounds good.

alya commented:
I'll let @amanagr do the next round of review here.

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -493,12 +543,16 @@ export function activate({
             task_data.handle_event(event.sender_id, event.data);
         }
 
         render_task_list_title();
```
Is there a reason we want to update task list title on every event update? Otherwise, this seems like a good opportunity to include that in `update_todo_widget` function.

pratik-pc replied:
No, it's better to update on the title event. I mostly left the title logic untouched earlier.

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -371,6 +455,21 @@ export function activate({
     }
 
+    function update_task_edit_controls(key: string): void {
+        const has_task =
```
Can you add comment for what this function does?

pratik-pc replied:
I added a comment and also changed the function name to make it more intuitive.

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -492,6 +654,8 @@ export function activate({
         const html = render_widgets_todo_widget_tasks(widget_data);
         $elem.find("ul.todo-widget").html(html);
         $elem.find(".widget-error").text("");
+        $elem.find("ul.todo-widget .todo-task-list-item-bar").toggle();
+        $elem.find("ul.todo-widget .todo-edit-task-list-item").toggle(is_my_task_list);
```
It's not clear just looking at this code what you want to achieve here. Either you should add comments or modify the code to be more clear.

pratik-pc replied:
This is bad way to hide the input elements, now that you mention. It is supposed to run once to hide input fields after loading the initial html. Changed it to use `.hide()`

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -507,21 +671,62 @@ export function activate({
-                    .append($(render_widgets_todo_widget_tasks({all_tasks: [task]})));
+            case "new_task": {
+                const $html = $(render_widgets_todo_widget_tasks({all_tasks: [task]}));
+                $html.find(".todo-task-list-item-bar").toggle();
```
Again, I think just using `toggle()` doesn't make it clear what's the exact behaviour that we want to achieve. So, this needs to be audited everywhere here.

amanagr commented:
@pratik-pc this looks great! Left some comments.

pratik-pc commented:
@amanagr addressed above feedbacks

amanagr commented:
Thanks @pratik-pc! Changes look good to me. Tested all possible event types manually too.

pratik-pc commented on web/src/todo_widget.ts:
```diff
@@ -471,6 +494,34 @@ export function activate({
+        if (!last_event || last_event.type === "new_task_list_title") {
+            render_task_list_title();
+            return;
+        }
```
I'll probably need to revisit this and improve a few things. I took a quick look at the overall code, and it will closes on message list re-render. I guess I could make it behave more like message edit and re-render the entire widget and set the state of the edit field.

@timabbott does it still make sense to re-render individual tasks for widget events, or should we switch to a single re-render logic for widgets?

timabbott replied:
I think for TODOs at least, which will never be huge, complete rerender seems fine. Polls may have other constraints, since you can imagine thousands of votes, perhaps? Not sure even there it'd be a constraint.

Honestly I kinda want us to try rewriting the TODO widget using modern JS framework components.

choucavalier commented:
@pratik-pc thanks for your work! do you plan on finishing up or should I take it from here?

pratik-pc committed:
todo_widget: Allow task name and description to be edited by author.

We add the ability to edit task name and description for todo widget
author. We store the edit state so that on any rerenders, we can
update the edit input fields.

Fixes part of zulip#20213.

pratik-pc committed:
todo_widget: Change selector for add task bar elements.

A later commit adds an inline edit bar to each todo task item whose
inputs reuse the `.add-task` and `.add-desc` classes already used by
the add-task bar at the bottom of the widget.  Scope the add-task
bar's queries and event handlers to `.add-task-bar` first, so they
keep targeting only that bar once the duplicate classes appear.

This is a no-op refactor today, since the add-task bar is currently
the only place those classes are used.

pratik-pc committed:
widgets: Add success and error handler option for widget server call.

We add option to perform success & error handling when executing
server callback.

amanagr commented on web/templates/widgets/todo_widget_tasks.hbs:
```diff
@@ -13,5 +13,15 @@
             </span>
         </label>
+        <i class="fa fa-pencil small todo-edit-task-list-item"></i>
+        <div class="todo-edit-task-list-item-bar">
```
The class names are not clear for what they are. Improving them would allow easy interpretation of JavaScript calls.

amanagr replied:
Also, we should make the edit buttons always visible on mobile otherwise mobile users won't be able to edit.

amanagr commented on web/src/generic_widget.ts:
```diff
@@ -97,10 +106,16 @@ export function render_widget_instance(info: {
             msg_type: "widget",
             data,
+            ...(on_success !== undefined && {on_success}),
+            ...(on_error !== undefined && {on_error}),
```
We don't need a spread here.

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -146,13 +150,63 @@ export function render({
+        const data = task_data.handle.edit_task.outbound(new_task, new_desc, key);
+        if (data) {
+            callback(data, success_handler, error_handler);
+        }
```
Should there be an `else` here to reset the state?

amanagr commented on web/src/todo_widget.ts:
```diff
@@ -146,13 +150,63 @@ export function render({
+        const task_exists = task_data.name_in_use(new_task);
+        const old_task_list_item = task_data.get_task_item(key);
+
+        assert(old_task_list_item);
```
We should return silently here in case user deleted the item in another window.

amanagr commented:
@pratik-pc left some comments, should be ready for integration after that.

pratik-pc commented:
@amanagr addressed above reviews

amanagr commented:
When I edit a task description and save, it says task already exists:

<img width="871" height="254" alt="Screenshot 2026-05-19 at 5 27 30 PM" src="https://github.com/user-attachments/assets/b29e8631-8793-495f-80f0-6a0a168a0951" />

amanagr commented:
Also, the location where the error is shown is incorrect. Can we show it somewhere near the task for which the error occured?

amanagr commented:
@pratik-pc I think this is close to completion, just needs a few final adjustments.

pratik-pc commented:
Oh, sorry missed the notification.

> Also, the location where the error is shown is incorrect. Can we show it somewhere near the task for which the error occured?

How should we handle this? Should we show the same error message on the individual task row, or perhaps display a tooltip on the save button?

amanagr commented:
>  the same error message on the individual task row

This seems right.

pratik-pc committed:
widgets: Hide widget-error DOM element when it is empty.

Hide `.widget-error` elements when they do not contain any text.

A later commit in this series gives each todo task item its own error
container for the inline edit bar.  Without this rule, those empty
containers would reserve vertical space and misalign the task rows, so
we collapse any error container that holds no text.

pratik-pc commented:
Changes:

- Fix the description edit error.
- Show individual error messages for each row.
- Reduce code duplication by introducing helper functions.
- Restore input focus and text selection across rerenders (e.g., a new task server event).
- First commit removes the space occupied by an empty `.widget-error` element.

amanagr commented:
@evykassirer can you review? (You were the only person who has worked in this area even though it was via TypeScript migration.)

evykassirer commented:
I might be missing earlier context, but I wouldn't expect the buttons to overflow like this, I think the line should shrink the inputs (to keep it all on one line) and eventually move both edit buttons to a second row

<img width="644" height="237" alt="image" src="https://github.com/user-attachments/assets/2e5bc764-424e-47a0-af53-ec01ac5c8484" />

I would also expect the space between edit inputs to line up with the "new task" inputs
